torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+kir::Expr* findVectorizedSet(
+    std::vector<kir::ForLoop*>& loop_structure,
+    const std::vector<kir::ForLoop*>& for_loops) {
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
+        (first_expr->isA<kir::UnaryOp>() &&
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set);
+    if (has_vectorize_op) {
+      fl->setVectorize(true);
+      loop_structure.push_back(fl);
+      return first_expr;
+    }
+  }
+  return nullptr;
+}
+
+// Get full extent for the inner-most, merged root domain
+kir::Val* getVectorizeExtent(kir::TensorIndex* node) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  auto domain = node->view()->domain()->rootDomain();
+  auto root_domain_index_map = node->rootDomainIndexMap();
+
+  auto extent = domain.back()->extent();
+  for (int64_t i = domain.size() - 2; i >= 0; --i) {
+    if (!root_domain_index_map[i]) {
+      extent = ir_builder.mulExpr(extent, domain[i]->extent());
+    } else {
+      break;
+    }
+  }
+  return extent;
+}
+
+// Generate Index Value
+kir::Val* generateBaseIndex(kir::TensorIndex* node) {
+  if (node->indices().size() == 1) {
+    return node->indices().front();
+  }
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  auto result = node->indices().front();
+  for (size_t idx = 1; idx < node->indices().size(); ++idx) {
+    result = ir_builder.addExpr(result, node->indices()[idx]);
+  }
+  return result;
+}
+
+kir::Val* setupNamedScalar(
+    kir::Scope& body,
+    kir::Val* val,
+    const std::string& name) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* one = ir_builder.create<kir::Int>(1);
+
+  auto namedScalar = ir_builder.namedSetExpr(name, val);
+  auto alloc =
+      ir_builder.create<kir::Allocate>(namedScalar, MemoryType::Local, one);
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
+
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // body -> allocate, read, compute, write for-loops
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* zero = ir_builder.create<kir::Int>(0);
+
+  // create new base For-Loop
+  const auto new_loop = ir_builder.create<kir::ForLoop>(
+      for_loop->index(), for_loop->extent(), for_loop->iter_domain());
+
+  // Find child For-Loops and add remaining expressions to base For-Loop
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  // Add child For-Loop to loop_structure
+  // Enable vectorize flag in child For-Loop
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+
+  // out_tv is the TensorView with the misaligned vec iterDomain
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
```

Comment:
Are we certain `vec_expr->outputs().front()` is always `kir::TensorView`? I don't find any explicit check.

Reply:
I added a check.